src/Streamly/Internal/Data/Fold/Type.hs

```diff
@@ -815,9 +967,42 @@ teeWithMin = undefined
 -- /Unimplemented/
 --
 {-# INLINE shortest #-}
-shortest :: -- Monad m =>
+shortest ::  Monad m =>
     Fold m x a -> Fold m x a -> Fold m x a
```

Comment:
We can make it more general by using:

```
shortest :: Fold m x a -> Fold m x b -> Fold m x (Either a b)
```

Same for `longest` as well.
```

Reply:
with longest: we get a type error:

longest :: Monad m => Fold m x a -> Fold m x b -> Fold m x (Either a b)
longest (Fold stepL beginL doneL) (Fold stepR beginR doneR) =
    Fold step begin done

    where
    begin :: Monad m => (m (Step (LongestRunner s1 s2 b1 b2) (Either a b)))
    begin = do
        resL <- beginL
        resR <- beginR
        return
            $ case resL of
                    Partial sL ->
                        Partial
                            $ case resR of
                                Partial sR -> LongestRunBoth sL sR
                                Done bR -> LongestRunLeft sL bR
                    Done bL -> first (LongestRunRight bL) resR

    step (LongestRunBoth sL sR) a = do
        resL <- stepL sL a
        resR <- stepR sR a
        case resL of
            Partial sL1 ->
                return
                    $ Partial
                    $ case resR of
                        Partial sR1 -> LongestRunBoth sL1 sR1
                        Done bR -> LongestRunLeft sL1 bR
            Done bL ->
                return
                    $ case resR of
                        Partial sR1 -> Partial $ LongestRunRight bL sR1
                        Done bR -> Done $ Right bR
    step (LongestRunLeft sL bR) a = do
        resL <- stepL sL a
        return
            $ case resL of
                Partial sL1 -> Partial $ LongestRunLeft sL1 bR
                Done bL -> Done $ Left bL
    step (LongestRunRight bL sR) a = do
        resR <- stepR sR a
        return
            $ case resR of
                Partial sR1 -> Partial $ LongestRunRight bL sR1
                Done bR -> Done $ Right bR

    done (LongestRunLeft sL _) = do
        x <- doneL sL
        return $ Left x
        --doneL sL

    done (LongestRunRight _ sR) = do
        x <- doneR sR
        return $ Right x
        --doneR sR
    --Since none has finished we can take the first one
    done (LongestRunBoth sL _)  = do
        x <- doneL sL
        return $ Left x
        --doneL sL

src/Streamly/Internal/Data/Fold/Type.hs:1017:9: error:
    • Couldn't match type ‘b’ with ‘Either a1 b4’
      ‘b’ is a rigid type variable bound by
        the type signature for:
          longest :: forall (m :: * -> *) x a b.
                     Monad m =>
                     Fold m x a -> Fold m x b -> Fold m x (Either a b)
        at src/Streamly/Internal/Data/Fold/Type.hs:1008:1-71
      Expected type: m1 (Step (LongestRunner s2 s3 b1 b2) (Either a1 b4))
        Actual type: m1 (Step (LongestRunner s s1 a b) b)